### ISSUE TITLE
Expression objects lack shape/iteration protocol, blocking array-style model code

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/discopt#816]
**Version:** discopt 0.6.1.dev0, editable install from a `main` working tree at
commit `adc54132` (with 3 uncommitted files, so the exact state is not
reproducible from the SHA alone). Python 3.13.5.

Note: `importlib.metadata.version('discopt')` reports `0.7.0` in that
environment, while `discopt.__version__` and `pyproject.toml` both say
`0.6.1.dev0` — stale dist-info left behind by an earlier editable install. Minor,
but it made pinning the version briefly confusing.

## Summary

Building a method-of-lines PDE model with `discopt.dae.DAEBuilder` currently
requires writing the right-hand side scalar-by-scalar, because expression
objects do not carry `.shape` and do not support NumPy-style array operations.
The same physics written once against `jax.numpy` cannot be reused to build
discopt expressions.

Separately — and more urgently — **`Variable` supports `__getitem__` but not
`__len__`/`__iter__`, and out-of-bounds indexing does not raise `IndexError`.
Any accidental iteration therefore hangs forever.** That one is a plain bug and
is probably worth fixing independently of the enhancement.

## Context

I am building a transient gas-network study that compares shooting (diffrax)
against full discretization. A hard requirement is that **both legs consume one
definition of the physics** — if the two legs' residuals ever drift, every
timing comparison in the study is void.

The shared residual is written against `jax.numpy` and operates on arrays:

```python
def momentum_residual(pipe, q_faces, p_points, eps):
    p_left, p_right = p_points[:-1], p_points[1:]
    dpdx = pipe.area_m2 * (p_right - p_left) / pipe.dx_m
    friction = pipe.K * q_faces * jnp.sqrt(q_faces**2 + eps**2) / (0.5*(p_left+p_right))
    return -(dpdx + friction)
```

I had hoped to pass `sqrt=discopt.sqrt` and reuse this verbatim for the
`DAEBuilder` leg. That is not currently possible.

## Bug: iteration hangs, out-of-bounds indexing is silent

```python
import discopt as d
m = d.Model()
x = m.continuous('x', shape=(5,), lb=0.1, ub=10.0)

hasattr(x, '__len__')      # False
hasattr(x, '__iter__')     # False
hasattr(x, '__getitem__')  # True

x[99]                      # returns IndexExpression -- does NOT raise IndexError
list(x)                    # hangs forever
```

Because `__getitem__` exists without `__iter__`, Python falls back to the legacy
iteration protocol, calling `x[0]`, `x[1]`, `x[2]`, … and stopping only on
`IndexError`. Since out-of-range indices return an expression instead, iteration
never terminates.

This bites on anything that iterates implicitly: `list(x)`, `for v in x`,
`sum(x)`, `np.array(x)`, `min`/`max`, unpacking. It cost me a wedged process
before I worked out what was happening, and the silent out-of-bounds indexing is
a correctness hazard in its own right — `x[99]` on a shape-`(5,)` variable
should not quietly build an expression.

Suggested fix, in rough priority order:

1. Raise `IndexError` for out-of-range scalar indices.
2. Add `__len__` returning `shape[0]`.
3. Add an explicit `__iter__` yielding the leading-axis elements (or, if
   iteration is deliberately unsupported, raise `TypeError` with a message
   pointing at the intended API).

## Enhancement: shape propagation and array ops on expressions

Measured behaviour today:

| expression | type | `.shape` |
|---|---|---|
| `x` | `Variable` | `(5,)` |
| `x[0]` | `IndexExpression` | absent |
| `x[:-1]` | `IndexExpression` | absent |
| `x[1:] - x[:-1]` | `BinaryOp` | absent |
| `d.sqrt(x*x + 0.01)` | `FunctionCall` | absent |
| `d.sqrt(x[:-1]*x[:-1] + 0.01)` | `FunctionCall` | absent |

The good news: elementwise arithmetic and `d.sqrt` **do** work on slices, and
compose. What is missing is the metadata to write generic array code — after one
operation the result no longer knows it is a length-4 vector, so anything that
branches on shape, or that needs `concatenate`/`stack`, has to be unrolled by
hand.

Concretely, this is what I want to be able to write:

```python
p_left, p_right = p_points[:-1], p_points[1:]        # shape (n,) each
dpdx = area * (p_right - p_left) / dx                 # shape (n,)
friction = K * q * sqrt(q*q + eps*eps) / (0.5*(p_left+p_right))
return -(dpdx + friction)                             # shape (n,)
```

and have it produce either JAX arrays or discopt expressions depending only on
which `sqrt` is passed in.

Requests, roughly in order of value to me:

1. **Propagate `.shape` through `IndexExpression`, `BinaryOp`, `FunctionCall`.**
   This alone unlocks most generic code.
2. **`concatenate` / `stack`** for building a vector from pieces — the
   method-of-lines pattern is "node value, interior values, node value", which
   currently has to be a Python loop.
3. **A documented `__array_ufunc__` / `__array_function__` path**, so
   `np.sqrt(expr)` dispatches to `discopt.sqrt`. That would make NumPy-generic
   model code work unmodified.

## Workaround I am using

I refactored the shared physics to expose a scalar, backend-agnostic form using
only `+ - * /` and `sqrt`:

```python
def momentum_rhs_scalar(pipe, q, p_left, p_right, eps, sqrt=jnp.sqrt):
    dpdx = pipe.area_m2 * (p_right - p_left) / pipe.dx_m
    friction = pipe.K * q * sqrt(q*q + eps*eps) / (0.5*(p_left+p_right))
    return -(dpdx + friction)
```

with a test asserting the scalar and vectorized forms agree to 1e-14. This works
and preserves the single-source-of-truth property, but the collocation leg then
has to loop over ~159 states per collocation point in Python to assemble the
RHS, rather than expressing it as a handful of array operations.

## Priority from my side

The iteration hang is the part I would fix first — it is small, and it turns a
typo into a wedged process. The shape-propagation enhancement is the one that
would change how I write models, but I have a workable path without it.

Happy to test a branch against this study's model, which is a reasonable
stress case: 159 states, 8 pipes, 2 compressors, index-1 DAE reduced to an ODE.
